### PR TITLE
ClusterLoader - Resources cleanup

### PR DESCRIPTION
**Pull request by krzysied**
Removing  resources based on instances in the test state.

### CODE REVIEW

File: clusterloader2/pkg/state/state.go [hunk 1]
```diff
@@ -95,3 +102,20 @@ func (ns *NamespacesState) Delete(namespace string, identifier InstancesIdentifi
 	delete(namespaceState, identifier)
 	return nil
 }
+
+// ListAll returns a list of all registered instances in namespaces.
+func (ns *NamespacesState) ListAll() []InstanceTuple {
```
Comment (shyamjvs) [inline clusterloader2/pkg/state/state.go]: Do you really need to transform map to list for what you're doing in the cleanup function?

Comment (krzysied) [reply clusterloader2/pkg/state/state.go]: Yes/no/depends.
Yes - I want to list namespace, identifier and instances without giving access to the map. Cause interaction with state is done only using methods, implementation and structure of state can be changed any time (as long method signatures remain the same [we can introduce interface here])
No - I can return whole map. This way I can remove all methods and operate directly on structure.

IMHO first solution is better cause it has better encapsulation.
We can think about other way of iterating over items like iterating over namespaces and then for every namespace iterate over identifiers and then for every namespace, identifier getting instances. However, it looks like list approach is simpler.

File: clusterloader2/pkg/test/simpletestexecutor.go [hunk 2]
```diff
@@ -226,8 +226,25 @@ func isErrsCritical(errList []error) bool {
 	return false
 }
 
-func cleanupResources(ctx Context) {
+func (ste *simpleTestExecutor) cleanupResources(ctx Context) {
+	instancesList := ctx.GetState().ListAll()
```
[the remark this answers is not in the document]
Comment (shyamjvs) [reply clusterloader2/pkg/test/simpletestexecutor.go]: We also need to think of non-namespaced resources (though I guess this is not very important atm) - if we should delete those automatically or rely on a config step for it.

[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/test/simpletestexecutor.go]: > I don't think you should be deleting all the created resources one by one as part of the cleanup function. IMO you should just delete the created namespaces (optionally waiting until they get deleted) and the namespace-controller will take care of deleting resources present in the namespace.

Resources can be created in already existing namespaces. These namespaces will not be removed by cluster loader.

>If user still wants to manually delete all resources from the test, they'll need to add another step in their config setting all replicas to 0. This way, it'll also be able to possible to delete based on a tuningSet.

That's one possible solution. However IMHO test should always (try to) leave cluster in exactly same state as it was at the beginning of test.

**Comment by wojtek-t** [comment]
@krzysied - is that something we really need?
Deleting namespace will delete all objects anyway. So I don't see it as something particularly useful.

Can we close this or there was some reason why you wanted it?

**Comment by krzysied** [comment]
> Deleting namespace will delete all objects anyway. So I don't see it as something particularly useful.

@wojtek-t In most cases what you wrote is true. However, I remember assumption that clusterloader should be able to use not only automanaged namespaces, but also any already existing ones. User namespaces will not be deleted by clusterloader! IMHO we should do the cleanup or ban custom namespaces.

**Comment by wojtek-t** [comment]
OK - yes that's true. We should do the cleanup.

**Comment by wojtek-t** [comment]
Closing - if this will be needed, we will reopen.

**Pull request closed without merging**